tests/integration_toolbox.py

```diff
@@ -156,6 +158,46 @@ def testGetUrlToFileCachedKeepaliveChanged(self):
             conn.close()
         self.assertEqual(b'This is a test\n', data)
 
+    def testGetUrl404(self):
+        """Call get_url for a file that doesn't exist"""
+        with self.assertRaises(HTTPError):
+            data = spacepy.toolbox.get_url(
+                'http://localhost:{}/doesnotexist.txt'.format(self.port))
+    
+    def testGetUrlNoConnNoKeepAlive(self):
+        """Call get_url without specifying connection or keepalive"""
+        ctype = http.client.HTTPConnection
+        connection = ctype("localhost:{}".format(self.port))
+        with self.assertRaises(ValueError):
+            spacepy.toolbox.get_url('http://localhost:{}/'.format(self.port), conn=connection, keepalive=False)
+    
+    def testGetUrlReadableWritable(self):
+        """Call get_url to trigger readable and writable"""
+        ctype = http.client.HTTPConnection
+        connection = ctype("localhost:{}".format(self.port))
+        connection.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM) 
+        _, connection = spacepy.toolbox.get_url('http://localhost:{}/'.format(self.port), conn=connection, keepalive=True)
+        self.assertEqual(connection.sock, None)
+    
+    def testGetUrlValueErrorOn404WithKeepAlive(self):
+        """Call get_url with keepalive and a bad link"""
+        ctype = http.client.HTTPConnection
+        connection = ctype("localhost:{}".format(self.port))
+        with self.assertRaises(RuntimeError):
+            spacepy.toolbox.get_url('http://localhost:{}/doesnotexist.txt'.format(self.port), keepalive=True)
+    
+    def testGetUrlCacheNoOutfileNoKeepAlive(self):
+        """Call get_url with Cache, no outfile, and no keepalive"""
+        ctype = http.client.HTTPConnection
+        connection = ctype("localhost:{}".format(self.port))
+        with self.assertRaises(ValueError):
```

Comment:
The ValueErrors at least should actually check the message in the exception, see e.g. testSparsePrevInsert in test_pycdf. I think most of those RuntimeErrors are also checkable.

Reply:
I made all of the new test cases check the message of the exceptions